src/controller/robot/controller.go

```diff
@@ -0,0 +1,343 @@
+	"github.com/goharbor/harbor/src/lib/errors"
+	"github.com/goharbor/harbor/src/lib/log"
+	"github.com/goharbor/harbor/src/lib/q"
+	"github.com/goharbor/harbor/src/pkg/permission/types"
+	"github.com/goharbor/harbor/src/pkg/project"
+	"github.com/goharbor/harbor/src/pkg/rbac"
+	rbac_model "github.com/goharbor/harbor/src/pkg/rbac/model"
+	robot "github.com/goharbor/harbor/src/pkg/robot2"
+	"github.com/goharbor/harbor/src/pkg/robot2/model"
+	"time"
+)
+
+var (
+	// Ctl is a global variable for the default robot account controller implementation
+	Ctl = NewController()
+)
+
+// Controller to handle the requests related with robot account
+type Controller interface {
+	// Get ...
+	Get(ctx context.Context, id int64, option *Option) (*Robot, error)
+
+	// Count returns the total count of robots according to the query
+	Count(ctx context.Context, query *q.Query) (total int64, err error)
+
+	// Create ...
+	Create(ctx context.Context, r *Robot) (int64, error)
+
+	// Delete ...
+	Delete(ctx context.Context, id int64) error
+
+	// Update ...
+	Update(ctx context.Context, r *Robot) error
+
+	// List ...
+	List(ctx context.Context, query *q.Query, option *Option) ([]*Robot, error)
+}
+
+// controller ...
+type controller struct {
+	robotMgr robot.Manager
+	proMgr   project.Manager
+	rbacMgr  rbac.Manager
+}
+
+// NewController ...
+func NewController() Controller {
+	return &controller{
+		robotMgr: robot.Mgr,
+		proMgr:   project.Mgr,
+		rbacMgr:  rbac.Mgr,
+	}
+}
+
+// Get ...
+func (d *controller) Get(ctx context.Context, id int64, option *Option) (*Robot, error) {
+	robot, err := d.robotMgr.Get(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	return d.populate(ctx, robot, option)
+}
+
+// Count ...
+func (d *controller) Count(ctx context.Context, query *q.Query) (int64, error) {
+	return d.robotMgr.Count(ctx, query)
+}
+
+// Create ...
+func (d *controller) Create(ctx context.Context, r *Robot) (int64, error) {
+	if err := d.setProjectID(ctx, r); err != nil {
+		return 0, err
+	}
+
+	if r.ExpiresAt == 0 {
+		tokenDuration := time.Duration(config.RobotTokenDuration()) * time.Minute
+		r.ExpiresAt = time.Now().UTC().Add(tokenDuration).Unix()
+	}
+
+	key, err := config.SecretKey()
+	if err != nil {
+		return 0, err
+	}
+	str := utils.GenerateRandomString()
+	secret, err := utils.ReversibleEncrypt(str, key)
+	if err != nil {
+		return 0, err
+	}
+
+	robotID, err := d.robotMgr.Create(ctx, &model.Robot{
+		Name:        r.Name,
+		Description: r.Description,
+		ProjectID:   r.ProjectID,
+		ExpiresAt:   r.ExpiresAt,
+		Secret:      secret,
+	})
+	if err != nil {
+		return 0, err
+	}
+
+	if err := d.createPermission(ctx, r); err != nil {
+		return 0, err
+	}
+	return robotID, nil
+}
+
+// Delete ...
+func (d *controller) Delete(ctx context.Context, id int64) error {
+	if err := d.robotMgr.Delete(ctx, id); err != nil {
+		return err
+	}
+	if err := d.rbacMgr.DeletePermissionsByRole(ctx, ROBOTTYPE, id); err != nil {
```

Comment:
DeletePermissionsByRole is not a proper function name, it seems it need to take the role or role_id as the parameter, deletePermissions is OK